torch/quantization/observer.py

```diff
@@ -137,7 +141,304 @@ def calculate_qparams(self):
+                beta = next_beta
+
+            if next_start_bin == start_bin and next_end_bin == end_bin:
+                continue
+
+            # calculate the quantization error using next_start_bin and next_end_bin
+            norm = self._compute_quantization_error(
+                next_start_bin, next_end_bin, norm_type
+            )
+
+            if norm > norm_min:
+                break
+            norm_min = norm
+            start_bin = next_start_bin
+            end_bin = next_end_bin
+
+        new_min = self.min_val + bin_width * start_bin
+        new_max = self.min_val + bin_width * (end_bin + 1)
+        return new_min, new_max
+
+    def _combine_histograms(
+        self, dst_histogram, dst_min, dst_max, src_histogram, src_min, src_max
+    ):
+        bins_dst = dst_histogram.size()[0]
+        bins_src = src_histogram.size()[0]
+
+        dst_bin_width = (dst_max - dst_min) / bins_dst
+        src_bin_width = (src_max - src_min) / bins_src
+
+        for i in range(bins_src):
+            src_bin_count = src_histogram[i].item()
+            if src_bin_count == 0:
+                continue
+
+            src_bin_begin = src_min + src_bin_width * i
+            src_bin_end = src_bin_begin + src_bin_width
+
+            dst_bin = 0
+            if dst_bin_width:
+                dst_bin = int((src_bin_begin - dst_min) / dst_bin_width)
+
+            dst_bin_begin = dst_min + dst_bin_width * dst_bin
+            dst_bin_end = dst_bin_begin + dst_bin_width
+
+            dst_bin2 = 0
+            if dst_bin_width:
+                dst_bin2 = min(
+                    int((src_bin_end - dst_min) / dst_bin_width), bins_dst - 1
+                )
+
+            assert dst_bin2 <= dst_bin + 2, "1 src_bin is mapped to at most 2 dst_bins"
+            # dst_bin_cnt is the count from src_bin that should go to dst_bin
+            # the remainder should go to dst_bin2
+            dst_bin_cnt = 0
+            if src_bin_width == 0 or dst_bin_width == 0:
+                dst_bin_cnt = src_bin_count
+            else:
+                # We divide counts in src_bin in proportion to range overlap with dst_bin
+                dst_bin_cnt = min(
+                    round(
+                        (dst_bin_end - src_bin_begin) / src_bin_width * src_bin_count
+                    ),
+                    src_bin_count,
+                )
+
+            dst_histogram[dst_bin] += dst_bin_cnt
+
+            # remaining should go to dst_bin2
+            if dst_bin_cnt < src_bin_count:
+                dst_histogram[dst_bin2] += src_bin_count - dst_bin_cnt
+
+    def forward(self, x):
+        with torch.no_grad():
+            min_val = self.min_val
+            max_val = self.max_val
+            histogram = self.histogram
+            if min_val is None or max_val is None or histogram is None:
+                min_val = torch.min(x)
+                max_val = torch.max(x)
+                if self.use_fixed_width is True:
```

Comment:
fixed_width is using the fixed width for the histogram, so each new input will be fit into the histogram whose width is the 2x range of the first batch. It's a less accurate way than combining histogram for each new input. It could be a little confusing to users, maybe better remove this option.